File: scripts/license.py
#!/usr/bin/env python3

import argparse
import glob
import re
import pathlib

license_header_rust = """
// Copyright 2025 Miku Push! Team
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
"""

license_header_typescript = """
/**
 * Copyright 2025 Miku Push! Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"""

print("adding license header to source code")

rust_source_files = glob.glob('src-tauri/**/*.rs', recursive=True) \
                    + glob.glob('lib/**/*.rs', recursive=True)

typescript_source_files = glob.glob('src/**/*.ts', recursive=True) \
                        + glob.glob('src/**/*.tsx', recursive=True) \
                        + glob.glob('src/**/*.css', recursive=True)

rust_regex = re.compile(r"^// Copyright 2025 Miku Push! Team(\n.*)+License.")
typescript_regex = re.compile(r"^/\*\*\n \* Copyright 2025 Miku Push! Team")

exclude_files = [
    "src/vite-env.d.ts",
    "src-tauri/build.rs",
    "src-tauri/src/main.rs",
    "src/components/ui/*.tsx",
    "lib/database/src/schema.rs"
]

def is_source_file_excluded(source_file_path: str):
    for exclude_file_pattern in exclude_files:
        path = pathlib.Path(source_file_path)

        if path.match(exclude_file_pattern):
            return True

    return False


def get_source_code(source_file_path: str):
    with open(source_file_path, "r", encoding="utf-8") as file:
        if is_source_file_excluded(source_file_path):
            print(f"excluding file {source_file_path}")
            return None

        return file.read().strip()


def add_license(source_files: list[str], license_header_content: str, regex: re.Pattern[str]):
    for source_file in source_files:
        source_code = get_source_code(source_file)
        if source_code is None:
            continue

        if regex.match(source_code):
            print(f"file {source_file} already has license header")
            continue

        with open(source_file, "w", encoding="utf-8") as file:
            file.write(license_header_content.strip() + "\n\n" + source_code)


def remove_license(source_files: list[str], license_header_content: str, regex: re.Pattern[str]):
    for source_file in source_files:
        source_code = get_source_code(source_file)
        if source_code is None:
            continue

        with open(source_file, "w", encoding="utf-8") as file:
            file.write(source_code.replace(license_header_content.strip(), "").strip())


parser = argparse.ArgumentParser(description='Add or remove license headers from source code files')
parser.add_argument(
    '--remove',
    action='store_true',
    help='Remove license headers'
)
args = parser.parse_args()

if args.remove:
    print("removing license header from source code")
    remove_license(rust_source_files, license_header_rust, rust_regex)
    remove_license(typescript_source_files, license_header_typescript, typescript_regex)
    print("removed license header from source code")
else:
    print("adding license header to source code")
    add_license(rust_source_files, license_header_rust, rust_regex)
    add_license(typescript_source_files, license_header_typescript, typescript_regex)
    print("added license header to all source code")

File: scripts/license.ts
import fs from 'node:fs'
import path from 'node:path'
import process from 'node:process'
import { globSync } from 'glob'
import ignore from 'ignore'

const licenseHeader = `
// Miku Push! is a simple, lightweight, and open-source WeTransfer alternative for desktop.
// Copyright (C) 2025  Miku Push! Team
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
`

const licenseHeaderTypescript = `
/**
 * Miku Push! is a simple, lightweight, and open-source WeTransfer alternative for desktop.
 * Copyright (C) 2025  Miku Push! Team
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
`

const rustSourcePatterns = [
  'src-tauri/**/*.rs',
  'lib/**/*.rs'
]

const swiftSourcePatterns = [
  'helpers/macos-helpers/Share/*.swift',
]

const typescriptSourcePatterns = [
  'src/**/*.ts',
  'src/**/*.tsx',
  'src/**/*.css'
]

const excludePatterns = [
  'src/vite-env.d.ts',
  'src-tauri/build.rs',
  'src-tauri/src/main.rs',
  'src/components/ui/*.tsx',
  'lib/database/src/schema.rs'
]

const ignoredFiles = ignore().add(excludePatterns)
const regex = /^\/\/ Miku Push! is a simple, lightweight, and open-source WeTransfer alternative for desktop\./
const typescriptRegex = /^\/\*\*\n \* Miku Push! is a simple, lightweight, and open-source WeTransfer alternative for desktop\./

const rootDir = process.cwd()

function collectFiles(patterns: string[]): string[] {
  const files = new Set<string>()

  patterns.forEach((pattern) => {
    const matches = globSync(pattern, {
      cwd: rootDir,
      nodir: true
    })

    matches
      .filter(match => !ignoredFiles.ignores(match))
      .forEach((match) => files.add(match))
  })

  return Array.from(files).sort()
}

function getSourceCode(filePath: string): string {
  const absolutePath = path.resolve(rootDir, filePath)
  const content = fs.readFileSync(absolutePath, 'utf-8')
  return content.trim()
}

function addLicense(sourceFiles: string[], licenseHeader: string, regex: RegExp): void {
  const trimmedHeader = licenseHeader.trim()

  for (const sourceFile of sourceFiles) {
    const sourceCode = getSourceCode(sourceFile)

    if (regex.test(sourceCode)) {
      console.log(`file ${sourceFile} already has license header`)
      continue
    }

    const absolutePath = path.resolve(rootDir, sourceFile)
    fs.writeFileSync(absolutePath, `${trimmedHeader}\n\n${sourceCode}`, 'utf-8')
    console.log('license header added on ', sourceFile)
  }
}

function removeLicense(sourceFiles: string[], licenseHeader: string): void {
  const trimmedHeader = licenseHeader.trim()

  for (const sourceFile of sourceFiles) {
    const sourceCode = getSourceCode(sourceFile)
    const absolutePath = path.resolve(rootDir, sourceFile)
    const updatedSource = sourceCode.replace(trimmedHeader, '').trim()
    fs.writeFileSync(absolutePath, updatedSource, 'utf-8')
    console.log('license header removed on ', sourceFile)
  }
}

const rustSourceFiles = collectFiles(rustSourcePatterns)
const swiftSourceFiles = collectFiles(swiftSourcePatterns)
const typescriptSourceFiles = collectFiles(typescriptSourcePatterns)

const args = process.argv.slice(2)
const shouldRemove = args.includes('--remove')

if (shouldRemove) {
  console.log('removing license header from source code')
  removeLicense(rustSourceFiles, licenseHeader)
  removeLicense(swiftSourceFiles, licenseHeader)
  removeLicense(typescriptSourceFiles, licenseHeaderTypescript)
  console.log('removed license header from source code')
} else {
  console.log('adding license header to source code')
  addLicense(rustSourceFiles, licenseHeader, regex)
  addLicense(swiftSourceFiles, licenseHeader, regex)
  addLicense(typescriptSourceFiles, licenseHeaderTypescript, typescriptRegex)
  console.log('added license header to all source code')
}
